### PR TITLE
list: refactor code, fix fish 2.3.0 globs

### LINKED ISSUE #322: bug: "No matches for wildcard" error on 'omf destroy'
I was doing a clean install of **omf** to verify the bug  report in #321. It failed as described there, and then I decided to remove omf so I could try again with modifications.

This was the result:
```fish
⋊> ~ omf destroy
This will destroy your Oh My Fish installation!
Are you sure you want to continue? (y/N) y
Removing Oh My Fish...
No matches for wildcard “$path/$name.*.copy”.
~/.local/share/omf/pkg/omf/functions/core/omf.destroy.fish (line 1): echo $path/$name.*.copy | tr ' ' '\n' | sort -r | head -1
                                                                          ^
in command substitution
        called on line 0 of file ~/.local/share/omf/pkg/omf/functions/core/omf.destroy.fish

in function “__omf.destroy.restore_backup”
        called on line 28 of file ~/.local/share/omf/pkg/omf/functions/core/omf.destroy.fish
        with parameter list “/home/oranja/.config/fish/functions/fish_prompt.fish”

in function “omf.destroy”
        called on line 7 of file ~/.local/share/omf/pkg/omf/functions/cli/omf.cli.destroy.fish

in function “omf.cli.destroy”
        called on line 38 of file ~/.local/share/omf/pkg/omf/functions/omf.fish

in function “omf”
        called on standard input
        with parameter list “destroy”

Welcome to fish, the friendly interactive shell
Type help for instructions on how to use fish
```

* For the record, this time it was _fish 2.3.0_, although I doubt it matters here.

### PR DESCRIPTION
**Pull request by derekstavis**
Fixes #322 and #321

### CODE REVIEW

File: pkg/omf/functions/packages/omf.packages.list.fish [hunk 1]
```diff
@@ -1,69 +1,85 @@
...
+    contains -i -- --available $argv
+    contains -i -- -a $argv
+    contains -i -- --installed $argv
+    contains -i -- -i $argv
+    contains -i -- --database $argv
+    contains -i -- -d $argv
+  end)
+
+  set option $argv[$option_index]
+  set type $argv[$type_index]
```
Comment (oranja) [inline pkg/omf/functions/packages/omf.packages.list.fish]: `-l` in these two lines too.

File: pkg/omf/functions/packages/omf.packages.list.fish [hunk 2]
```diff
@@ -1,69 +1,85 @@
...
-  __omf.packages.sort $list
+  __omf.packages.format $list
 end
 
 function __omf.packages.list.installed -a type
   set -l list
+  set -l installed_package_path $OMF_PATH/pkg/*
+  set -l installed_theme_path $OMF_PATH/themes/*
+
+  echo $type
```
Comment (oranja) [inline pkg/omf/functions/packages/omf.packages.list.fish]: Seems like a rouge `echo`

File: pkg/omf/functions/packages/omf.packages.list.fish [hunk 3]
```diff
@@ -1,69 +1,85 @@
-function __omf.packages.sort
-  for package in (echo $argv | tr ' ' '\n' | sort); echo $package; end
-end
-
-function __omf.packages.builtin
-  echo "omf"
-  echo "fish-spec"
-end
+set -l builtin_package_path {$OMF_PATH,$OMF_CONFIG}/pkg*/{omf,fish-spec}
```
Comment (oranja) [inline pkg/omf/functions/packages/omf.packages.list.fish]: Only used in `__omf.packages.list.installed`. Nest it there?

File: pkg/omf/functions/packages/omf.packages.list.fish [hunk 4]
```diff
@@ -1,69 +1,85 @@
...
-  test "$type" = "--theme"; or for package in (basename -a $OMF_PATH/db/pkg/*)
-    contains $package (basename -a {$OMF_CONFIG,$OMF_PATH}/pkg/*); or set list $list $package
+  test "$type" = "--theme"; or for path in $database_package_path
+    contains $path $installed_package_path; or set list $list $path
   end
 
-  test "$type" = "--plugin"; or for package in (basename -a $OMF_PATH/db/themes/*)
-    contains $package (basename -a {$OMF_CONFIG,$OMF_PATH}/themes/*); or set list $list $package
+  test "$type" = "--plugin"; or for path in $database_theme_path
+    contains $path $installed_theme_path; or set list $list $path
```
Comment (oranja) [inline pkg/omf/functions/packages/omf.packages.list.fish]: Here and in the previous `test` block, it's comparing two full paths from globbing two different folders. Need to `__omf.packages.format` the lists beforehand.

**Comment by oranja** [comment]
Aside from the comments above, I confirm that this allowed me to install Oh-My-Fish on fish 2.3.0 (fixes #321).
```fish
$ rm -rf ~/{.local/share,.config}/omf
$ curl -L http://get.oh-my.fish -o omf_install.fish
$ env OMF_REPO_BRANCH=refactor-list fish omf_install.fish
Installing Oh My Fish to /home/oranja/.local/share/omf...
Cloning refactor-list from https://github.com/oh-my-fish/oh-my-fish.git...
Existent config.fish found at /home/oranja/.config/fish
↳ Moving to /home/oranja/.config/fish/config.1464505350.copy
Adding startup code to fish config file...
Building Oh My Fish configuration...
Installing theme default
✔ theme default successfully installed.
Installation successful!
Welcome to fish, the friendly interactive shell
Type help for instructions on how to use fish
```

There's still a problem with `destroy`, though:
```fish
No matches for wildcard “$path/$name.*.copy”.
~/.local/share/omf/pkg/omf/functions/core/omf.destroy.fish (line 1): echo $path/$name.*.copy | tr ' ' '\n' | sort -r | head -1
                                                                          ^
in command substitution
        called on line 0 of file ~/.local/share/omf/pkg/omf/functions/core/omf.destroy.fish

in function “__omf.destroy.restore_backup”
        called on line 28 of file ~/.local/share/omf/pkg/omf/functions/core/omf.destroy.fish
        with parameter list “/home/oranja/.config/fish/functions/fish_prompt.fish”

in function “omf.destroy”
        called on line 7 of file ~/.local/share/omf/pkg/omf/functions/cli/omf.cli.destroy.fish

in function “omf.cli.destroy”
        called on line 38 of file ~/.local/share/omf/pkg/omf/functions/omf.fish

in function “omf”
        called on standard input
        with parameter list “destroy”
```

File: pkg/omf/functions/packages/omf.packages.list.fish [hunk 5]
```diff
@@ -72,6 +88,6 @@ function omf.packages.list -a option type
   case "-i" "--installed"
     __omf.packages.list.installed $type
   case "*"
-    __omf.packages.list $type
+    __omf.packages.list.installed $type
```
Comment (bobthecow) [inline pkg/omf/functions/packages/omf.packages.list.fish]: if this is the case, couldn't we just get rid of the case before (or add `"*"` to the matching case and combine the two)?

**Pull request merged**